### PR TITLE
ORTModule support non-differentiable module output

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Not all the module outputs are differentiable. For example, some module will output the mask, shape, size, or indices. We can't build the gradient graphs for such module outputs, and we don't need to feed any gradient tensor either. As a result, Yield op need to allow a mismatching number of inputs and outputs. 

There are a few changes in this PR: 
- When building gradient graph, we first check if all the module outputs are differentiable. If not, they are marked as "non_differentiable" and excluded from the back-propagation. 
- Introduced a new attribute "non_differentiable_outputs" to indicate which module outputs doesn't need a gradient.
- Added  a unit test to cover the non-differentiable  output case. 
- Discovered two bugs: CastFromBoolToInt/CastFromBoolToFloat gives wrong result; ORTModule forward() is giving wrong result in one model.

### CODE REVIEW

**Comment by ytaous** [comment]
works for the ORTModule(encoder) now - FYI

File: orttraining/orttraining/core/framework/gradient_graph_builder.cc [hunk 1]
```diff
@@ -51,19 +54,25 @@ GradientGraphBuilder::GradientGraphBuilder(Graph* graph,
...
 
     const Node* node = graph_->GetProducerNode(name);
     if (!node) {
       ORT_THROW(name, " couldn't find the producer node.");
     }
-    y_nodes_.insert(node);
+
+    if (forward_reachable_nodes.find(node) == forward_reachable_nodes.end()) {
+      non_differentiable_y_node_arg_names_.insert(name);
+      LOGS(logger_, WARNING) << "The model weigths and inputs are non-differentiable from " << name << ". "
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/core/framework/gradient_graph_builder.cc:65]:
```suggestion
      LOGS(logger_, INFO) << "The model weights and inputs are non-differentiable from " << name << ". "
```
Should it be INFO? We are trying to suppress all messages that the user cannot do anything about it during training. Currently implementation is verbose

nit: typo

File: orttraining/orttraining/python/training/ortmodule.py [hunk 2]
```diff
@@ -208,6 +208,10 @@ def backward(ctx, *grad_outputs):
                     # Push user output grads to ONNX backend.
                     contiguous_grad_outputs = []
                     for idx, grad_output in enumerate(grad_outputs):
+                        if idx in self._onnx_graphs_info.output_grad_indices_non_differentiable:
+                            assert grad_output is None
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:212]: Coudl you provide a meaningful error message here? This is a user-facing error, so a hint on the issue might help them to have insights on the root cause

File: orttraining/orttraining/core/framework/gradient_graph_builder.h [hunk 3]
```diff
@@ -88,6 +88,10 @@ class GradientGraphBuilder {
 
   Status Build(const std::unordered_set<std::string>* p_initializer_names_to_preserve = nullptr);
 
+  std::unordered_set<std::string> GetNonDifferentiableYNodeArgNames() const {
```
Comment (mrry) [inline orttraining/orttraining/core/framework/gradient_graph_builder.h:91]: Performance nit: return `const std::unordered_set<std::string>&` (and change call-site to assign the return value to a reference)?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 4]
```diff
@@ -154,6 +154,26 @@ def forward(self, x, y=None, z=None):
...
+        out1 = self.relu(out)
+        out2 = self.fc2(out1)
+        mask1 = torch.gt(out1, 0.01)
+        mask1 = mask1.long()    # TODO: Casting from bool to float or int will cause the UT failure
+                                # True is casted to 1065353216 for Cast(from=bool, to=int), whereas pytorch would give 1
+                                # True is casted to -1 for Cast(from=bool, to=float), where as pytorch would give 1.0f
+        mask2 = torch.lt(out2, 0.02)
+        mask2 = mask2.long()
+        
+        return out1, mask1, out2, mask2     # intentionally place the non-differentialble output in the middle
```
Comment (mrry) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:175]: Typo: "non-differentiable"

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 5]
```diff
@@ -154,6 +154,26 @@ def forward(self, x, y=None, z=None):
...
+        self.relu = torch.nn.ReLU()
+        self.fc2 = torch.nn.Linear(hidden_size, num_classes)
+
+    def forward(self, input1):
+        out = self.fc1(input1)
+        out1 = self.relu(out)
+        out2 = self.fc2(out1)
+        mask1 = torch.gt(out1, 0.01)
+        mask1 = mask1.long()    # TODO: Casting from bool to float or int will cause the UT failure
+                                # True is casted to 1065353216 for Cast(from=bool, to=int), whereas pytorch would give 1
```
Comment (mrry) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:170]: Interesting! Is this another bug we should raise with the exporter team?

**Review by mrry — CHANGES_REQUESTED** [review]
Looks good... just a few nits.

**Comment by SherlockNoMad** [comment]
> works for the ORTModule(encoder) now - FYI

Thanks a lot for the verification!

File: orttraining/orttraining/core/framework/gradient_graph_builder.cc (hunk 1 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/core/framework/gradient_graph_builder.cc:65]: done.

File: orttraining/orttraining/training_ops/cpu/controlflow/yield.cc [hunk 6]
```diff
@@ -32,12 +32,18 @@ Status YieldOp::Compute(OpKernelContext* ctx) const {
     ORT_THROW("Terminating backward run, since the terminate is set to true.");
   } else {
     ORT_ENFORCE(backward_inputs.second.size() == static_cast<size_t>(ctx->OutputCount()));
-    for (int i = 0; i < ctx->OutputCount(); ++i) {
-      if (std::find(full_shape_outputs_.begin(), full_shape_outputs_.end(), static_cast<int64_t>(i)) !=
-          full_shape_outputs_.end()) {
-        ORT_ENFORCE(ctx->Input<Tensor>(i)->Shape() == backward_inputs.second[i].Get<Tensor>().Shape());
+    ORT_ENFORCE(ctx->InputCount() == ctx->OutputCount() + static_cast<int>(non_differentiable_outputs_.size()));
```
[the remark this answers is not in the document]
Comment (SherlockNoMad) [reply orttraining/orttraining/training_ops/cpu/controlflow/yield.cc:35]: this is actually already checked in the schema verification. I am deleting this line.

[the remark this answers is not in the document]
Comment (SherlockNoMad) [reply orttraining/orttraining/training_ops/cpu/controlflow/yield.cc:35]: this line 
https://github.com/microsoft/onnxruntime/pull/7048/files#diff-836b7a7261cac41b48c2fe26d6565d47591669eb5f8b627d255da6b940ea3c8cR2234

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 5 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:170]: I haven't look into the details on the root cause of this bug... Will keep you posted :)

File: orttraining/orttraining/core/framework/gradient_graph_builder.h [hunk 7]
```diff
@@ -118,9 +124,16 @@ class GradientGraphBuilder {
   // key: name of the gradient, value: num of gradients pending
   std::unordered_map<std::string, int> pending_;
 
+  /**
+  Perferms a BFS on the graph
```
[the remark this answers is not in the document]
Comment (SherlockNoMad) [reply orttraining/orttraining/core/framework/gradient_graph_builder.h:128]: good suggestion. updated.

File: orttraining/orttraining/core/framework/gradient_graph_builder.h (hunk 3 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/core/framework/gradient_graph_builder.h:91]: done.

File: orttraining/orttraining/python/training/ortmodule.py (hunk 2 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/python/training/ortmodule.py:212]: done.

File: orttraining/orttraining/training_ops/cpu/controlflow/yield.cc [hunk 8]
```diff
@@ -32,12 +32,18 @@ Status YieldOp::Compute(OpKernelContext* ctx) const {
...
+
+    for (int i = 0, j = 0; i < ctx->InputCount(); ++i) {
+      if (non_differentiable_outputs_.count(static_cast<int64_t>(i)) > 0) {
```
[the remark this answers is not in the document]
Comment (SherlockNoMad) [reply orttraining/orttraining/training_ops/cpu/controlflow/yield.cc:38]: done.

**Review by mrry — APPROVED** [review]

File: orttraining/orttraining/python/training/ortmodule.py [hunk 9]
```diff
@@ -208,6 +208,12 @@ def backward(ctx, *grad_outputs):
                     # Push user output grads to ONNX backend.
                     contiguous_grad_outputs = []
                     for idx, grad_output in enumerate(grad_outputs):
+                        if idx in self._onnx_graphs_info.output_grad_indices_non_differentiable:
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule.py:211]: Is it possible to push this validation to the backend?

`self._onnx_graphs_info.output_grad_indices_non_differentiable` is populated in the backend and `grad_outputs` is passed in too

Comment (SherlockNoMad) [reply orttraining/orttraining/python/training/ortmodule.py:211]: This is actually a defensive programming practice...  
 
if grad_output is provided on non-differential-able outputs, the backend would crash on multiple places, e.g. the Yield's kernel assert on #gradient  == # Yield's output 
 
But I feel the best practice there to fail early, and provide the exact error message to user...

**Review by thiagocrepaldi — APPROVED** [review]
